CorsixTH/Lua/window.lua

```diff
@@ -113,6 +113,11 @@ function Window:onChangeResolution()
   if self.x_original and self.y_original then
     self:setPosition(self.x_original, self.y_original)
   end
+  if self.resize_redraw then
+    -- Reopen dialog in the new resolution in case the need for columns changes
+    self:close()
+    self.ui:addWindow(self.resize_redraw(self.ui))
```

Comment:
Big sledge hammer approach, move a road a bit by leveling the entire block of houses, and rebuilding them all.

It works, in a very crude way. For example, if you have multiple windows on the screen, this will break the layering order (what window is above which other windows) for the user. May not be a problem currently in the program.

Ignoring the big hammer, there are some parts not quite right here.

The first thing is `self.resize_draw`. You define it in the derived class, and use it in the base class. That is against object-oriented thinking, where the base class should never know anything about the derived class. (The other way around is allowed, although should be limited and/or run through functions to make code changes more easily). Here, the `Window` class should thus not know about the `resize_redraw` variable since the variable "lives" in a derived class.

If you want this approach (although I will argue below you probably don't), then `Window` should own `resize_redraw`. Derived classes may then access or change it. A minor secondary point is that there are better ways to retrieve the class (did you ever read the `class.lua` file comments?) from `self`. I'll leave that for another time, ask me if you really want to know.

The second thing is that the base class is in no position to truthfully re-create a new instance with all relevant derived class information preserved. It should NOT know about content of derived classes, so if the recreated window needs some of its information to be preserved in the next instance (eg scrollbar positions, buttons that are pushed or disabled, currently displayed data, etc), this approach will completely fail.

The only way to handle that is not do this here, and do it in the derived class instead. In other words, ask or tell the derived class to do this by means of a function call to it. The derived class knows what needs to be copied to the new instance. This also avoids breaking the construction process, in case a derived class has additional parameters in its constructor (beyond `self.ui`) that you now completely ignore.

Usually you'd create a new function for that. In this particular case, I am not sure that is needed. Is it an option to override `onChangeResolution` in the derived class instead?  Likely you'd want to call `Window.onChangeResolution(self)` there at some point to get the base class code executed.